monai/data/wsi_reader.py

```diff
@@ -940,6 +1054,91 @@ def get_mpp(self, wsi, level: int) -> tuple[float, float]:
 
         raise ValueError("`mpp` cannot be obtained for this file. Please use `level` instead.")
 
+    def get_wsi_at_mpp(self, wsi, mpp: tuple, atol: float = 0.00, rtol: float = 0.05) -> np.ndarray:
```

Comment:
This code appears to duplicate a lot of that in the other `get_wsi_at_mpp` method. Can this be factored out into a separate function to eliminate the duplication?

Reply:
Hi Eric, I factored out the actual resizing of the wsi to a given mpp resolution into a function called _resize_to_mpp_res. The tolerance checks still happen in the main function get_wsi_at_mpp.

Reply:
I'm looking at this method vs the one implemented in `CuCIMWSIReader` and it seems identical (or virtually so) except for the last line before return. What I'd suggest is that this method be moved to `BaseWSIReader` and then the implementation here and in `CuCIMWSIReader` calls it and converts its output to a numpy or cupy array as appropriate.

Reply:
Based on the idea to factor out the function to compute the target resolution, we could also factor out the tolerance checks into a BaseWSIReader function. That would again reduce redundancy, since they are indeed identical.

Reply:
If you could please do this that would greatly improve things, thanks.

Reply:
Hi Eric, this is now implemented. There is now a function _compute_mpp_tolerances to check tolerances in BaseWSIReader which is called in every reader's get_wsi_at_mpp function. I think the latter is now minimized as much as possible. Still need to finish the test functions though.

Reply:
Looks good now thanks. If you're still working on tests I don't need to review it all quite yet.